### PR TITLE
Added functionality to save a reservation

### PR DESCRIPTION
**Pull request by Faka-GLB**
Created a singleton class to serve as database to save reservations and the number of parking lots, also added a fields check to a reservation before saving it.

image

project tree

### CODE REVIEW

File: app/src/main/java/com/onboarding/parkingsystemjava/activity/MainActivity.java [hunk 1]
```diff
@@ -32,4 +33,5 @@ private void setListeners() {
     public void sendInput(int input) {
         presenter.setParkingLots(input);
     }
+
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/activity/MainActivity.java]: remove this extraline pls

File: app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java [hunk 2]
```diff
@@ -31,11 +34,13 @@ public void setListeners() {
     }
 
     @Override
-    public void sendStartDateTime(String dateTime) {
-        presenter.setStartDateTextView(dateTime);
+    public void sendStartDateTime(String dateTimeString, Calendar dateTimeCalendar) {
+        presenter.setStartDateTextView(dateTimeString);
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java]: I think that you can unify this calls

File: app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java [hunk 3]
```diff
@@ -31,11 +34,13 @@ public void setListeners() {
...
+        presenter.setReservationStartDate(dateTimeCalendar);
     }
 
-    public void sendEndDateTime(String dateTime) {
-        presenter.setEndDateTextView(dateTime);
+    public void sendEndDateTime(String dateTimeString, Calendar dateTimeCalendar) {
+        presenter.setEndDateTextView(dateTimeString);
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java]: same here

File: app/src/main/java/com/onboarding/parkingsystemjava/database/ParkingDatabase.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package com.onboarding.parkingsystemjava.database;
+
+import com.onboarding.parkingsystemjava.entity.Reservation;
+import java.util.ArrayList;
+
+public class ParkingDatabase {
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/database/ParkingDatabase.java:6]: excellent!

File: app/src/main/java/com/onboarding/parkingsystemjava/database/ParkingDatabase.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    private static ParkingDatabase single_instance = null;
+    private final ArrayList<Reservation> reservations = new ArrayList<>();
+    private int parkingLots;
+
+    private ParkingDatabase() {
+    }
+
+    public static ParkingDatabase getInstance()
+    {
+        if (single_instance == null)
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/database/ParkingDatabase.java]: change the name to singleInstance, to follow the naming convention pls

File: app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Calendar;
+
+public class Reservation {
+    private Calendar startDate;
+    private Calendar endDate;
+    private int parkingLot;
+    private String userPassword;
+
+    public Reservation(){
+        this.parkingLot=-1;
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java]: put this -1 into a constant pls and remember to format the code properly (ctrl + alt + l)

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ParkingPresenter.java [hunk 7]
```diff
@@ -1,5 +1,6 @@
 package com.onboarding.parkingsystemjava.mvp.presenter;
 
+import com.onboarding.parkingsystemjava.entity.Reservation;
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ParkingPresenter.java]: are you using this import?

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveComprobation.java [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package com.onboarding.parkingsystemjava.mvp.model;
+
+public enum ReserveComprobation {
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveComprobation.java]: pls move this to another package as this is not a model. Maybe a util package

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java [hunk 9]
```diff
@@ -22,7 +28,15 @@ public void onEndDateButtonPress() {
 
     @Override
     public void onOkButtonPress() {
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java:31]: you can unify these two methods here
```
model.setParkingLot(view.getParkingLot());
model.setUserPassword(view.getUserPassword());
```
into a one like
`model.saveReservationInfo(view.getParkingLot(), view.getUserPassword())`

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java [hunk 10]
```diff
@@ -39,8 +41,40 @@ public void setEndDateTextView(String date) {
...
-        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
-        getActivity().startActivity(intent);
+        getActivity().finish();
+    }
+
+    @Override
+    public int getParkingLot() {
+        String lot = binding.editTextNewReservationActivityLotNumber.getText().toString();
+        if (lot.isEmpty())
+            return -1; //-1 used in model to check if user has set a parking lot
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java]: do not hardcode pls, use constants

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java [hunk 11]
```diff
@@ -39,8 +41,40 @@ public void setEndDateTextView(String date) {
...
+    }
+
+    @Override
+    public String getUserPassword() {
+        return binding.editTextNewReservationActivityPassword.getText().toString();
+    }
+
+    @Override
+    public void showComprobationToast(ReserveComprobation comprobation) {
+        switch (comprobation) {
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java]: this switch should not be in the view, should be in the presenter, and remember to check that the getActivity is not null before using it

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java (hunk 2 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/activity/MainActivity.java (hunk 1 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/activity/MainActivity.java]: removed

File: app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java (hunk 3 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/database/ParkingDatabase.java (hunk 5 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/database/ParkingDatabase.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ParkingPresenter.java (hunk 7 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ParkingPresenter.java]: nope, removed

File: app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java (hunk 6 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java (hunk 9 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java:31]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java (hunk 10 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java]: switched to constant

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveComprobation.java (hunk 8 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveComprobation.java]: moved to utils package

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java (hunk 11 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java]: done

**Review by lucianoeli — APPROVED** [review]
Great job!

File: app/src/main/java/com/onboarding/parkingsystemjava/database/ParkingDatabase.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    private static ParkingDatabase singleInstance = null;
+    private final ArrayList<Reservation> reservations = new ArrayList<>();
+    private int parkingLots;
+
+    private ParkingDatabase() {
+    }
+
+    public static ParkingDatabase getInstance()
+    {
+        if (singleInstance == null)
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/database/ParkingDatabase.java]: please use the curly brackets here

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    public void setEndDate(Calendar endDate) {
+        this.reservation.setEndDate(endDate);
+    }
+
+    public ReserveComprobation reserveFieldsCheck(Reservation reservation) {
+        if (reservation.getStartDate() == null)
+            return MISSING_START;
+        else if (reservation.getEndDate() == null)
+            return MISSING_END;
+        else if (reservation.getParkingLot() == -1)
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: use constants pls

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+            return MISSING_LOT;
+        else if (reservation.getUserPassword().isEmpty())
+            return MISSING_PASSWORD;
+        else
+            return COMPROBATION_OK;
```
Comment (FranciscoGaraventa) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: You can change this, please use switch statement to grant code clarity.

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java [hunk 15]
```diff
@@ -22,19 +28,44 @@ public void onEndDateButtonPress() {
...
 
     @Override
     public void onCancelButtonPress() {
         view.returnToMainActivity();
     }
 
-    public void setStartDateTextView(String date) {
+    public void setReservationStartDate(Calendar startDate, String date) {
+        model.setStartDate(startDate);
         view.setStartDateTextView(date);
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java]: here you should use the info from the model
`        view.setStartDateTextView(model.getStartDate());`

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java [hunk 16]
```diff
@@ -22,19 +28,44 @@ public void onEndDateButtonPress() {
...
     }
 
-    public void setEndDateTextView(String date) {
+    public void setReservationEndDate(Calendar endDate, String date) {
+        model.setEndDate(endDate);
         view.setEndDateTextView(date);
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java]: same here

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java [hunk 17]
```diff
@@ -39,8 +40,53 @@ public void setEndDateTextView(String date) {
...
-        getActivity().startActivity(intent);
+        Activity activity = getActivity();
+        if (activity != null)
+            getActivity().finish();
+    }
+
+    @Override
+    public int getParkingLot() {
+        String lot = binding.editTextNewReservationActivityLotNumber.getText().toString();
+        if (lot.isEmpty())
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java]: use the curly brackets

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java [hunk 18]
```diff
@@ -39,8 +40,53 @@ public void setEndDateTextView(String date) {
...
+            return - ConstantUtils.INT_ONE; //-1 used in model to check if user has set a parking lot
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java]: create a MINUS_ONE_INT constant instead of this pls

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java (hunk 14 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: switch requires primitive types to evaluate, and I'm evaluating a Reservation, which has two Calendars, an int and a String, so java doesn't let me use a switch and recommend use an if statement.

File: app/src/main/java/com/onboarding/parkingsystemjava/database/ParkingDatabase.java (hunk 12 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/database/ParkingDatabase.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java (hunk 13 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java (hunk 17 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java (hunk 18 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/view/ReserveView.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java (hunk 15 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java (hunk 16 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/presenter/ReservePresenter.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java (hunk 14 above)
Comment (FranciscoGaraventa) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: Okey, so this implementation is fine.

**Review by SartiMau — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

**Pull request merged**